### PR TITLE
Support storing application level info with authenticated peer's principal

### PR DESCRIPTION
**Pull request by sophokles73**

### CODE REVIEW

**Comment by sophokles73** [comment]
@boaks please wait with merging, I need to check something first ...

**Comment by sophokles73** [comment]
@boaks ok, all set. I forgot to commit my latest changes ;-)
@sbernard31 does this work for you as well?

File: scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/ServerHandshaker.java [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2015, 2018 Institute for Pervasive Computing, ETH Zurich and others.
+ * Copyright (c) 2015, 2019 Institute for Pervasive Computing, ETH Zurich and others.
```
Comment (sbernard31) [inline scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/ServerHandshaker.java]: I feel this class should not be modified but you find a unused import so why not :)

File: scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/Handshaker.java [hunk 2]
```diff
@@ -984,6 +991,7 @@ protected final void handshakeStarted() throws HandshakeException {
 	}
 
 	protected final void sessionEstablished() throws HandshakeException {
+		amendPeerPrincipal();
```
Comment (sbernard31) [inline scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/Handshaker.java]: I don't kown if `amendPeerPrincipal` shoud be in or out the `if block` :thinking:

File: scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/Handshaker.java [hunk 3]
```diff
@@ -63,10 +63,12 @@
 import java.security.GeneralSecurityException;
 import java.security.MessageDigest;
 import java.security.NoSuchAlgorithmException;
+import java.security.Principal;
 import java.security.PrivateKey;
 import java.security.PublicKey;
 import java.security.cert.X509Certificate;
 import java.util.ArrayList;
+import java.util.Collections;
```
Comment (boaks) [inline scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/Handshaker.java]: Not used.

File: scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/Handshaker.java (hunk 2 above)
Comment (boaks) [inline scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/Handshaker.java]: I assume, that `amendPeerPrincipal` should only be called once per handshake. To ensure that, consider to move it after

```
if (!sessionEstablished) {
	sessionEstablished = true;
```

**Comment by boaks** [comment]
Is there a intended use-case, which requires the public `Principal`'s constructors with the `AdditionalInfo` parameter? If not, I would remove them. Beside of that and the "small" comments above, LGTM.

File: scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/Handshaker.java (hunk 2 above)
Comment (sophokles73) [reply scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/Handshaker.java]: just out of curiosity: under what circumstances will *sessionEstablished* be already be `true` when this method is invoked?

**Comment by sophokles73** [comment]
> Is there a intended use-case, which requires the public Principal's constructors with the AdditionalInfo parameter? If not, I would remove them.

No, there isn't. Thanks for pointing this out. Will make them private ...

File: scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/Handshaker.java (hunk 2 above)
Comment (boaks) [reply scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/Handshaker.java]: Your right. It was introduced to ensure, that this doesn't happen accidentally (may be in the future).
May be we convert that into an exception, but after 2.0.0-M16 .

**Comment by boaks** [comment]
The most could be removed, only the one private with the most parameter is required.
For `PreSharedKeyIdentity` you may consider to remove the second private one (that with the additional "name").

File: scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/Handshaker.java (hunk 2 above)
Comment (sophokles73) [reply scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/Handshaker.java]: Does this mean that I should keep the method invocation at its current position?

Comment (boaks) [reply scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/Handshaker.java]: If you want, you could keep it there. 
I would still prefer to move it.

Comment (sophokles73) [reply scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/Handshaker.java]: ok, will do

**Pull request merged**